rsc/src/main/java/com/cloudera/livy/rsc/rpc/Rpc.java

```diff
@@ -275,10 +273,14 @@ public void operationComplete(ChannelFuture cf) {
       };
 
       dispatcher.registerRpc(id, promise, msg.getClass().getName());
-      synchronized (channelLock) {
```

Comment:
You are totally right. I weighed the size of solution 1 and 2 and I think 1 is smaller but in the long term, 2 is better.